src/operations/DocumentHighlight.cpp

```diff
@@ -0,0 +1,294 @@
+
+    auto [locations, kinds] = findPropertyReferencesWithKinds(source, property, ty, module->astTypes);
+    highlights.reserve(locations.size());
+
+    for (size_t i = 0; i < locations.size(); ++i)
+    {
+        highlights.emplace_back(createHighlight(textDocument, locations[i], kinds[i]));
+    }
+
+    return highlights;
+}
+
+// Determines whether the name matches a type reference in one of the provided generics
+// If so, we find the usages inside of that node
+static bool handleIfTypeReferenceByName(Luau::AstNode* node, Luau::AstArray<Luau::AstGenericType> generics,
+    Luau::AstArray<Luau::AstGenericTypePack> genericPacks, Luau::AstName name, std::vector<lsp::DocumentHighlight>& highlights,
+    const TextDocument* textDocument)
+{
+    if (!isTypeReference(name, generics, genericPacks))
+        return false;
+
+    auto [locations, kinds] = findTypeParameterUsagesWithKinds(*node, name);
+
+    highlights.reserve(highlights.size() + locations.size());
+
+    for (size_t i = 0; i < locations.size(); ++i)
+    {
+        highlights.push_back(createHighlight(*textDocument, locations[i], kinds[i]));
+    }
+
+    return true;
+}
+
+// Determines whether the name matches a type reference in one of the provided generics
+// If so, we find the usages inside of that node
+static bool handleIfTypeReferenceByPosition(Luau::AstNode* node, Luau::AstArray<Luau::AstGenericType> generics,
+    Luau::AstArray<Luau::AstGenericTypePack> genericPacks, Luau::Position position, std::vector<lsp::DocumentHighlight>& highlights,
+    const TextDocument* textDocument)
+{
+    auto name = findTypeReferenceName(position, generics, genericPacks);
+    if (!name)
+        return false;
+
+    auto [locations, kinds] = findTypeParameterUsagesWithKinds(*node, name.value());
+
+    highlights.reserve(highlights.size() + locations.size());
+
+    for (size_t i = 0; i < locations.size(); ++i)
+    {
+        highlights.push_back(createHighlight(*textDocument, locations[i], kinds[i]));
+    }
+
+    return true;
+}
+
+lsp::DocumentHighlightResult WorkspaceFolder::documentHighlight(const lsp::DocumentHighlightParams& params)
+{
+    auto moduleName = fileResolver.getModuleName(params.textDocument.uri);
+    auto textDocument = fileResolver.getTextDocument(params.textDocument.uri);
+    if (!textDocument)
+        throw JsonRpcException(lsp::ErrorCode::RequestFailed, "No managed text document for " + params.textDocument.uri.toString());
+    auto position = textDocument->convertPosition(params.position);
+
+    // Run the type checker to ensure we are up to date
+    // We check for autocomplete here since autocomplete has stricter types
+    checkStrict(moduleName);
+
+    auto sourceModule = frontend.getSourceModule(moduleName);
+    if (!sourceModule)
+        throw JsonRpcException(lsp::ErrorCode::RequestFailed, "Unable to read source code");
+
+    auto exprOrLocal = findExprOrLocalAtPositionClosed(*sourceModule, position);
```

Comment:
I think document highlight should process only the current file for performance reasons, if you mean that it could process like the references request but only use the results for the current file. If I remember correctly, there's nothing preventing us from having a more unified setup, so I can implement that.